File: Tests/LedgelingsTests/ShippedCreaturesTests.swift
import Foundation
import LedgelingsCore
import Testing
@testable import Ledgelings

/// The built-in creatures are written in the letter format under sprites/text/.
/// This keeps the shipped sheets in step with them, and rebuilds them when asked:
///
///     LEDGELINGS_BUILD_CREATURES=1 swift test --filter ShippedCreaturesTests
@Suite struct ShippedCreaturesTests {
    static let repo = URL(fileURLWithPath: #filePath).deletingLastPathComponent().deletingLastPathComponent().deletingLastPathComponent()
    static let texts = repo.appendingPathComponent("sprites/text")
    static let resources = repo.appendingPathComponent("Sources/Ledgelings/Resources/sprites")
    static let building = ProcessInfo.processInfo.environment["LEDGELINGS_BUILD_CREATURES"] == "1"

    @Test(arguments: SpriteLibrary.builtIn.filter { $0 != "blocky" })
    func shippedSheetMatchesItsText(name: String) throws {
        let text = try String(contentsOf: Self.texts.appendingPathComponent("\(name).txt"), encoding: .utf8)
        let sheet = try SpriteText.parse(text)
        #expect(sheet.name == name)
        #expect(sheet.kind != nil && sheet.cast.count >= 3, "a built-in creature knows what it is and has a cast")
        let image = SpriteText.pixels(sheet, palette: .blocky)
        if Self.building {
            try SpriteLibrary.writePNG(image, to: Self.resources.appendingPathComponent("\(name).png"))
            let meta = SpriteText.atlas(name: name, kind: sheet.kind, cast: sheet.cast)
            try JSONSerialization.data(withJSONObject: meta, options: [.prettyPrinted, .sortedKeys])
                .write(to: Self.resources.appendingPathComponent("\(name).json"))
            return
        }
        let shipped = try SpriteAtlas(named: name)
        #expect(try shipped.image() == image, "\(name).png differs from sprites/text/\(name).txt; rebuild with LEDGELINGS_BUILD_CREATURES=1")
        #expect(shipped.meta.kind == sheet.kind)
        #expect(shipped.meta.cast == sheet.cast)
    }
}

File: win/Ledgelings.Tests/ShippedCreaturesTests.cs
namespace Ledgelings.Tests;

/// <summary>The built-in creatures are written in the letter format under sprites/text/ at the
/// repo root. This keeps the shipped sheets in step with them. (The rebuild path the macOS
/// suite offers under LEDGELINGS_BUILD_CREATURES is not ported; run `swift test` for that.)</summary>
public class ShippedCreaturesTests
{
    /// <summary>The repo root: the first ancestor of the test binary that holds sprites/text.</summary>
    private static string Repo
    {
        get
        {
            for (var dir = new DirectoryInfo(AppContext.BaseDirectory); dir is not null; dir = dir.Parent)
                if (Directory.Exists(Path.Combine(dir.FullName, "sprites", "text"))) return dir.FullName;
            throw new DirectoryNotFoundException("sprites/text not found above " + AppContext.BaseDirectory);
        }
    }

    public static IEnumerable<object[]> TextCreatures => SpriteLibrary.BuiltIn.Where(n => n != "blocky").Select(n => new object[] { n });

    [Theory]
    [MemberData(nameof(TextCreatures))]
    public void ShippedSheetMatchesItsText(string name)
    {
        var text = File.ReadAllText(Path.Combine(Repo, "sprites", "text", name + ".txt"));
        var sheet = SpriteText.Parse(text);
        Assert.Equal(name, sheet.Name);
        Assert.True(sheet.Kind is not null && sheet.Cast.Count >= 3, "a built-in creature knows what it is and has a cast");
        var image = SpriteText.Pixels(sheet, SpriteText.Palette.Blocky);
        var shipped = SpriteAtlas.Named(name).Image();
        var why = $"{name}.png differs from sprites/text/{name}.txt; rebuild with LEDGELINGS_BUILD_CREATURES=1 swift test";
        Assert.True(shipped.Width == image.Width && shipped.Height == image.Height, why);
        Assert.True(shipped.Rgba.AsSpan().SequenceEqual(image.Rgba), why);
        var meta = SpriteAtlas.Named(name).Info;
        Assert.Equal(sheet.Kind, meta.Kind);
        Assert.Equal(sheet.Cast, meta.Cast);
        Assert.Equal(sheet.Colour?.Hex, meta.Colour);
    }
}
